File: 00994. Rotting Oranges/solution.js
/**
 * Problem: 994. Rotting Oranges
 *
 * Difficulty: Medium
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

/**
 * Finds minimum time for all oranges to rot, or -1 if impossible
 *
 * @param {number[][]} grid - Grid of oranges (0=empty, 1=fresh, 2=rotten)
 *
 * @returns {number} - Minutes required or -1
 */
const orangesRotting = (grid) => {
  // Track count of fresh oranges
  let freshCount = 0

  // Get grid dimensions and initialize queue for BFS
  const rows = grid.length,
    cols = grid[0].length,
    rottenQueue = []

  // Find all rotten oranges and count fresh oranges
  for (let i = 0; i < rows; i++)
    for (let j = 0; j < cols; j++)
      if (grid[i][j] === 2) rottenQueue.push([i, j])
      else if (grid[i][j] === 1) freshCount++

  // No fresh oranges means no time needed
  if (freshCount === 0) return 0
  // No rotten oranges means impossible to rot all fresh oranges
  if (rottenQueue.length === 0) return -1

  // Helper function to check and convert fresh orange to rotten
  const rotOrange = (i, j) => {
    if (i < 0 || j < 0 || i >= rows || j >= cols || grid[i][j] !== 1)
      return false

    grid[i][j] = 2
    freshCount--

    return true
  }

  // Queue for next minute's oranges to process
  const nextMinuteQueue = []
  // Track elapsed time
  let minutes = 0

  // BFS to rot oranges level by level (by minute)
  while (rottenQueue.length > 0 || nextMinuteQueue.length > 0) {
    // If current minute queue is empty, move to next minute
    if (rottenQueue.length === 0 && nextMinuteQueue.length > 0) {
      minutes++

      // Move all oranges from next minute to current queue
      while (nextMinuteQueue.length > 0)
        rottenQueue.push(nextMinuteQueue.shift())
    }

    // Process current rotten orange
    if (rottenQueue.length > 0) {
      const [i, j] = rottenQueue.shift()

      // Check all 4 adjacent cells
      if (rotOrange(i + 1, j)) nextMinuteQueue.push([i + 1, j])
      if (rotOrange(i, j + 1)) nextMinuteQueue.push([i, j + 1])
      if (rotOrange(i - 1, j)) nextMinuteQueue.push([i - 1, j])
      if (rotOrange(i, j - 1)) nextMinuteQueue.push([i, j - 1])
    }
  }

  // Return minutes if all oranges rotted, otherwise -1
  return freshCount === 0 ? minutes : -1
}

File: 00994. Rotting Oranges/solution.go
/**
 * Problem: 994. Rotting Oranges
 *
 * Difficulty: Medium
 *
 * Language: Go
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

// Four possible movement directions: right, left, down, up
var directions = [4][2]int{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}

const (
	Fresh  = 1
	Rotten = 2
)

func orangesRotting(grid [][]int) int {
	// Get grid dimensions
	rows, cols := len(grid), len(grid[0])
	// Queue to store positions of rotten oranges
	var rottenQueue [][2]int

	// Count fresh oranges and collect initial rotten oranges
	freshCount := 0
	for i := 0; i < rows; i++ {
		for j := 0; j < cols; j++ {
			switch grid[i][j] {
			case Fresh:
				freshCount++
			case Rotten:
				rottenQueue = append(rottenQueue, [2]int{i, j})
			}
		}
	}

	// No fresh oranges means no time needed
	if freshCount == 0 {
		return 0
	}

	// BFS to rot oranges minute by minute
	minutes := 0
	for len(rottenQueue) > 0 {
		// Process all currently rotten oranges
		for _, orange := range rottenQueue {
			for _, dir := range directions {
				// Calculate adjacent cell position
				newRow, newCol := orange[0]+dir[0], orange[1]+dir[1]

				// If adjacent cell is valid and has a fresh orange, rot it
				if newRow >= 0 && newRow < rows && newCol >= 0 && newCol < cols && grid[newRow][newCol] == Fresh {
					grid[newRow][newCol] = Rotten
					rottenQueue = append(rottenQueue, [2]int{newRow, newCol})
					freshCount--
				}
			}

			// Remove processed orange from queue
			rottenQueue = rottenQueue[1:]
		}

		// Increment time after processing all oranges at current minute
		minutes++
		
		// If all oranges are rotten, return elapsed time
		if freshCount == 0 {
			return minutes
		}
	}

	// Return -1 if some oranges never rot
	return -1
}